packages/query-core/src/types.ts

```diff
@@ -14,7 +14,9 @@ export type QueryFunction<
   T = unknown,
   TQueryKey extends QueryKey = QueryKey,
   TPageParam = any,
-> = (context: QueryFunctionContext<TQueryKey, TPageParam>) => T | Promise<T>
+> = (
+  context: QueryFunctionContext<TQueryKey, TPageParam>,
+) => unknown extends T ? unknown : undefined extends T ? never : T | Promise<T>
```

Comment:
we also have a `NonUndefinedGuard` in the repo that looks a bit simpler:

https://github.com/TanStack/query/blob/3f057a24778c15c55665d11fc8415200a20e4840/packages/react-query/src/queryOptions.ts#L13

Reply:
This actually has a different meaning, it _only_ prevents the value `undefined` by itself. `undefined extends T` passes if the generic union contains `undefined`. 


I think they'll both correctly cause compile errors, but `undefined extends T` seems like it gives a clearer error message closer to the source. 

`T extends undefined`
![image](https://github.com/TanStack/query/assets/5288805/a477c9f8-a18d-499e-8de1-edb392b18b29)
`undefined extends T`
![image](https://github.com/TanStack/query/assets/5288805/7235502a-7a81-4a2f-93cd-2c8d8ea6bf3c)
 


See: https://www.typescriptlang.org/play?#code/C4TwDgpgBAKgogD2BAdgEwM4FV0QGYCWKEaAPDAHxQC8sUESqmUArrocWlAPxTEBuEAE5QAXLADcAKFCQoADRqxEydNnZESpFCwC2AI2FQAPqw2cK0gPRWodgHrcpM8NBxp8mtCqYYY5Klo2Dw4SekY1Ol4BI3EYaVloAE0ld09OHzV-HQMjU2D0kkspGzsoRyA

But curious for your thoughts! (maybe we should be switching that existing implementation too?)

Reply:
> maybe we should be switching that existing implementation too?

yeah that sounds like the right thing to do; I'll merge your PR and maybe you can drop another one that does this ? we have good test coverage for `initialData` typings

Reply:
hmm can you take a look at the errors now after I merged in beta here? they are not on the beta branch ...

Reply:
Yep, more than happy to follow up with the other change. 

Taking a look at the build failures! Might be a bit before I can figure out exactly what's wrong here. Something with `QueryPersister` isn't playing nice, I think from https://github.com/TanStack/query/pull/5125.

Reply:
Okay @TkDodo it should be hopefully fixed with 4eba53bd6ee78cc2fe50fe7223d5b2fc506aaf3d -- also added tests about `null` return values + initialState. 

Couple of notes:
- I honestly don't understand how the fix worked...

```ts
unknown extends T ? unknown : undefined extends T ? never : T | Promise<T>
// vs
unknown extends T ? T : undefined extends T ? never : T | Promise<T>
```
- I think it made the error messages slightly worse 

![image](https://github.com/TanStack/query/assets/5288805/d4b52538-4c60-4f81-9dfd-d86cf8d727ea)
- It also means that `data` ends up being inferred with the types, including `undefined` -- instead of `unknown` as it was previously. (because of this, I removed the type `Expect<>` and just left the `@ts-expect-error` annotations

![image](https://github.com/TanStack/query/assets/5288805/6700b35a-caf0-4b04-b380-97e92a1e6e4c)




Regardless of all of this weirdness, I still 100% think this is better than the existing state + avoids some nasty/unexpected runtime issues. But would also appreciate another quick smoketest from your side too!

Reply:
wow, this looks really good. Thank you 🙏